src/digest.c

```diff
@@ -72,16 +72,80 @@ FILE* open_with_widechar_on_windows(const char* txt) {
     return out;
 }
 
+
+// Also already used in sha2.h
+//
+// We can rely on WORDS_BIGENDIAN only be defined on big endian systems thanks to Rconfig.
+//
+// A number of other #define based tests are in other source files here for different hash
+// algorithm implementations notably crc32c, pmurhash, sha2 and xxhash
+//
+// A small and elegant test is also in package qs based on https://stackoverflow.com/a/1001373
+
+// edd 02 Dec 2013  use Rconfig.h to define BYTE_ORDER, unless already defined
+#ifndef BYTE_ORDER
+// see sha2.c comments, and on the internet at large
+#define LITTLE_ENDIAN 1234
+#define BIG_ENDIAN    4321
+#ifdef WORDS_BIGENDIAN
+#define BYTE_ORDER  BIG_ENDIAN
+#else
+#define BYTE_ORDER  LITTLE_ENDIAN
+#endif
+#endif
+
+SEXP is_big_endian(void) {
+  return Rf_ScalarLogical(BYTE_ORDER == BIG_ENDIAN);
+}
+
+SEXP is_little_endian(void) {
+  return Rf_ScalarLogical(BYTE_ORDER == LITTLE_ENDIAN);
+}
+
+/* shorthand memcpy for consistent re-use */
+#define MCP(WHAT) memcpy(output, (WHAT), output_length)
+/* for any algorithm that has output as unsigned char*, can use resolve */
+#define resolve(WHAT) if (leaveRaw) {                          \
+  memcpy(output, (WHAT), output_length);                       \
+} else for (int j = 0; j < output_length; j++) {               \
+  snprintf(output + j * 2, 3, "%02x", (WHAT)[j]);              \
+}                                                              \
+
+#define decl_tmp32(WHAT) uint32_t tmp = (WHAT);
+#define decl_tmp64(WHAT) uint64_t tmp = (WHAT);
+
+/* for any algorithm that has output as an integral value, can use dump */
+
+void rev_memcpy(char *dst, const void *src, int len) {
+  for (int i = 0; i < len; i++) {
+    dst[i] = ((char*)src)[len - i - 1];
+  }
+}
+
+#define dump32(WHAT) decl_tmp32(WHAT);     \
+if (leaveRaw) {                            \
+  rev_memcpy(output, &tmp, output_length);     \
+} else snprintf(output, 128, "%08x", tmp);
+
+#define dump64(WHAT) decl_tmp64(WHAT);     \
+if (leaveRaw) {                            \
+  rev_memcpy(output, &tmp, output_length);                                                     \
+} else snprintf(output, 128, "%016" PRIx64, tmp);
+
```

Comment:
Works for me. Happy to retool them as local functions instead of macros, but that is definitely getting into larger refactor territory.

How about this: keep these changes approximately as-is. Introduce new issue for discussion laying out some refactoring objectives (e.g. the repetitious cases for files vs not also annoyed me), and I'll contribute another PR against that.

Re macros - for some copy pasta, I'd likely use macros in C++ as well, but these are more a byproduct of how I thought through the approach (and then didn't wrap up as functions) rather than being a particularly good answer for this approach.

Reply:
Sounds good.  We have a mix of using helpers, and of repetition as the code is (in some places) thirty years old, we now know some things better than we did then ... and also because I was generally really prudent (you could say paranoid) about not changing working code once the package got more widely used.

Reply:
Alright, I will be on the lookout for new issue (feel free to tag / assign me / whatever). I'll discuss there a bit to reach consensus on scope / intent, then make a new PR.

Please let me know here any specific revisions needed for *this* change set here. Alternatively, should be setup for any maintainers of this repo to be able to make changes directly to the source of this PR, so y'all can just make changes directly.

Reply:
WILCO.

Aside: for the standard contributions content, y'all should should consider a CONTRIBUTING.md, explaining those details. Also, introducing issue / PR templates with those items (+whatever else). Pre/auto-nagging is the best kind.

Reply:
Sure.  But keep in mind that over the years I added dozens of repos here and it is tedious to copy this around.  I wrote things up once in my most-visible repo [in this file](https://github.com/RcppCore/Rcpp/blob/master/Contributing.md) and sometimes refer to it from other repos.  It's mostly about the 'discuss first, pr later' approach that can help (but that I myself also do not _religiously_ follow because somethings small things are just ... small).  [ Need to update that file to w.r.t. to C++ standards. ]

So while point taken, templates can also be heavy for some more marginal repos.  But we do this at work too etc and I know where you are coming from.  We had not needed it here yet.  (Also note above about paranoia / care.  I don't think I bulk formatted here either.)

[ And then as a Wilco (the band) fan and Chicagoan, where did that come from?  Love it.  :heart:  ]

Reply:
Release 0.6.37 is now at CRAN so I just pushed. It's mostly DESCRIPTION and the now mandatory Authors@R (to which you should also add yourself, likely in last spot (as the ctb list is roughly chronological)).

Reply:
yeah, heard re maintaining consistent infrastructure docs. github organizations make that relatively easy, with inherit / override as appropriate on a per project basis, but that kind of reorg is its own set of churn.

I'll update the PR to be against 0.6.37 & make the other changes (just did whitespace + ChangeLog, will get the others in a tic).

While I am a fan of Wilco and Chi-town, WILCO is a NATO standard phone talker / radio operator procedure word (short for "understood: will comply"). I also almost wrote just there "proword", which is the shorthand for procedure word.

Reply:
Today I learned ...